### DISCUSSION TITLE
Can one migrate to the new 2.0 type system incrementally without relying on the deprecated/alpha mypy plugin?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10949] [Usage Questions]
I am in the progress of trying to upgrade a very large code base from 1.4 to 2.0 and am running into thousands of type errors, of the form:

```
[...] has incompatible type "Optional[str]"; expected "str"
Incompatible types in assignment (expression has type "Decimal", variable has type "Union[SQLCoreOperations[Optional[_N]], _N, None]") 
Incompatible types in assignment (expression has type "Decimal", variable has type "Union[SQLCoreOperations[Optional[_N]], _N, None]") 
```
(among many others of a similar form - basically all stemming from the old ORM syntax and not using this updated syntax: https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping).

I would vastly prefer to upgrade in two separate steps - first upgrade the library, then fix the types. This will make it much easier to rollback the library upgrades if anything goes wrong. Less chance of conflicts and fewer reviews required. 

It's my understanding that the deprecated mypy plugin is the only option to upgrade incrementally, from the docs:

>  Existing code that passes Mypy checks using the plugin with SQLAlchemy 1.4 installed will continue to pass all checks in SQLAlchemy 2.0 without any changes required, provided the plugin is still used. SQLAlchemy 2.0’s API is fully backwards compatible with the SQLAlchemy 1.4 API and Mypy plugin behavior

However, I am having various issues getting it to work and am concerned the lack of support will lead me to a dead end. For example, I asked a question about it here: https://github.com/sqlalchemy/sqlalchemy2-stubs/issues/264 and the response was "don't use it".  I also this comment in a stack overflow thread:

> There is a [deprecated Mypy plugin](https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html) that can be used in the interim, but I found it unstable (several months ago) so I gave up on it and refactored my code. 

My question is, is this plugin indeed the only way to incrementally upgrade to SqlAlchemy 2.0? Is there any other simpler solution that would allow me to upgrade the libraries without simultaneously having to make thousands of syntax updates in order to get mypy to pass?

**Comment by zzzeek** [comment]
The mypy plugin in 2.0 should not have any regressions, so if you are using all 1.4 style constructs, mypy plugin in 2.0 should continue to work fine.

the problem will be if you begin using 2.0-only constructs like explicit Mapped[] calls, etc.  and apparently the DeclarativeBase class, the mypy plugin in theory should work smoothly with these things also, but if it doesnt, I'd advise using TYPE_CHECKING blocks to get through the interim.  

from my understanding, explicit use of Mapped[] in mapping setups should not be rejected by the mypy plugin, those lines should be ignored.

for the Base aspect of it, if you want mypy plugin to continue working while runtime uses DeclarativeBase, do this:

```py
if TYPE_CHECKING:
     Base = declarative_base()
else:
    class Base(DeclarativeBase):
        pass
```


the presence of `declarative_base()` is the main thing that sets the mypy plugin into action on a class.  otherwise it isn't going to take effect for a particular class.   you can work with TYPE_CHECKING blocks to incrementally move classes to use 2.0 declarative style while other classes are still on the old style, based on which `Base` class is seen within TYPE_CHECKING.

**Comment by ghost** [comment]
Thank you for your answer. 

Digging more into my issues with my plugin (since, This is the way), I'm still using the `sqlmypy` plugin that belongs to `SQLAlchemy-stubs` while on 1.4. In order to complete this migration incrementally, I believe I'll have to upgrade to `SQLaclchemy2-stubs` first. H

However I hit a brick wall when attempting this due to this issue: https://github.com/sqlalchemy/sqlalchemy2-stubs/issues/264, and was basically told, too bad, you can't use the plugin.

**Reply by zzzeek** [reply]
oh....the old dropbox stubs....yah we dont have any compatibility with those at all, sorry

**Reply by zzzeek** [reply]
all of these pre-2.0 stubs were all "interim hack" sorts of things.  none of these approaches were robust at all, so, to the degree you can relax your typing until you can get things working 2.0 style, will likely help.  in 2.0 we are treating typing as a top level robust thing, and we have some major improvements to it in 2.1 as well.

**Reply by ghost** [reply]
Ah, I think I understand where I went wrong. I should be able to keep using `sqlmypy` plus `sqlalchemy-stubs` for the partial migration. Had to throw in a few `# type : ignore`s but that's perfectly fine, as long as it isn't thousands of them!

Thank you for your help!!

**Comment by ghost** [comment]
I was able to upgrade to SqlAlchemy 2.0 but I am now looking at a way to incrementally update types to the new version.

My current setup is with SqlAlchemy 2.0, using the `sqlmypy` mypy plugin with `sqlalchemy-stubs` package installed. A few #type ignores are scattered about to get SqlAlchemy 2.0 to work. It's been in my backlog to remove these stubs and plugin, and I'm finally getting some time to work on this. I'm a bit nervous of our current setup since the documentation says all stub packages should be uninstalled when using 2.0.  Somehow we are using stubs and it seems to be working so far.

@zzzeek - I am rereading the conversation from a few months ago above, and thinking that perhaps there was a misunderstanding. You said "oh....the old dropbox stubs....yah we dont have any compatibility with those at all", but now that I look closer, the dropbox stubs are the `sqlalchemy-stubs` and my issue was with using `sqlalchemy2-stubs`. So now I'm thinking I perhaps should _not_ have continued using `sqlalchemy-stubs` and continued to try to get `sqlalchemy2-stubs` to work, but it didn't seem like any help would be provided for that.

It's unclear to me what next steps to take to upgrade incrementally to the 2.0 types. I have tried:

- Remove `sqlalchemy-stubs` and sqlmypy plugin = 30k type errors
- Remove `sqlalchemy-stubs` and use the sqlalchemy.ext.mypy.plugin mypy plugin = 9k type errors

I can't use any sqlalchemy 2.0 style things while the stubs are installed because I get errors like "there is no `mapped_column` in sqlalchemy.orm" - I guess the stubs package hides this legit type.

I've thought about moving `sqlalchemy-stubs` into my package as stubs via the mypy path and then adding types for things like `mapped_column` incrementally but I can't figure out how to do that in a type safe way - I'd have to guess at the real type for `mapped_column` and I'm not sure what it's supposed to look like. (there is no mapped_column in sqlalchemy2-stubs - thought maybe I could use it from that)

Any direction or help would be much appreciated. I am very eager to get onto the proper type system and out of this middle-ground hacky place. 😞

**Comment by ghost** [comment]
Actually looks like I'm out of luck - I'd have to actually downgrade to SqlAlchemy 1.4, since I hit this issue now:

> mypy.ini:8: error: Error importing plugin "sqlalchemy.ext.mypy.plugin": The SQLAlchemy mypy plugin in SQLAlchemy 2.0 does not work with sqlalchemy-stubs or sqlalchemy2-stubs installed, as well as with any other third party SQLAlchemy stubs.  Please uninstall all SQLAlchemy stubs packages.  [misc]

😭

**Comment by ghost** [comment]
Sorry, just a heads up that I created a duplicate issue here: https://github.com/sqlalchemy/sqlalchemy/discussions/11235

**Comment by zzzeek** [comment]
step 0:  "9000 typing errors" is mostly meaningless, since a typing error in one line of code that is referenced elsewhere will be multiplied out by how often it's referenced, which can then cascade out into references of that code, etc.   you could have one or two typing issues at the very base of your program that when fixed could resolve all 9000 errors in one step (it's unlikely to be *that* simple, but "9000 errors" is not by itself much indication of the scale of the task).

step 1:   do not use sqlalchemy-stubs or the `sqlmypy` extension that is specific to those stubs.    (also dont use sqlalchemy-stubs).
step 2:  *do* use the `sqlalchemy.ext.mypy.plugin` , with *both* version 1.4 and version 2.0 of SQLAlchemy, if you have 1.4 style mappings in use (those that use `Column`, not `mapped_column'`).
step 3:  if using SQLAlchemy 1.4, use sqlalchemy2-stubs.   if using SQLAlchemy 2.0, do **not** use **any** stubs.  SQLAlchemy 2.0 is fully typed inline.   **the plugin continues to work in either case** 
step 4:  begin your migration by turning down mypy's strictness.  enable mypy only for those .py files that contain model definitions first.   Fix those.    See https://mypy.readthedocs.io/en/stable/existing_code.html#existing-code for the basic strategy how to do this with mypy.    Migrate your class models to 2.0 style as documented [here](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping).   Fix for specific mypy warnings only, then continue to enable more mypy warnings to be fixed.
step 5: when all your models use `Mapped[]` and `mapped_column()` as documented at https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping , take out the `sqlalchemy.ext.mypy.plugin` entirely.

from that point on, there should be little to no SQLAlchemy-specific typing errors left.   Queries using ORM `Query` as well as Core/ORM `select()` are typed automatically.  it's basically just the models that need to be updated.

**Comment by ghost** [comment]
Thank you for your continued help! I'll break down some of the 9k errors.

**Need type annotation for [a column]**

Show up for tables where the columns are defined like this:

```
    creation_order_id = Column(BigInteger().with_variant(Integer(), "sqlite"), primary_key=True, autoincrement=True)
    foo_sid = Column(String(34), nullable=False)
    is_bar = Column(Boolean, nullable=False)
```

Basically I guess any column where I don't add an explicit type.

**Incompatible types in assignment**
 seems to show up when I have a column with an explicit type, defined like this:

```
    status: Column[ALiteralStatusType] = Column(
        Enum(*StringLiteralUtils.get_literal_type_values(ALiteralStatusType)), nullable=False
    )
```

**has incompatible type "str | None"; expected "str"**

Seems to show up even though the column has nullable false. Perhaps would go away if I explicitly set the type.
```
my_id_column = Column(String(36), nullable=False, unique=True)
 ```
 
 
If these column definitions are the only things causing errors, I think I could add them all incrementally by having the type defined in a common file:

sql_utils.py:
```
from sqlalchemy import Column as Column

Mapped = Column
```

and changing all table definitions to:
```
from mycode.sql_utils import Mapped

...
  my_id_column: Mapped[str] = Column(String(36), nullable=False, unique=True)
  is_bar: Mapped[bool] = Column(Boolean, nullable=False)
```

Then once every table definition has the type explicitly set, I can remove the old mypy plugin & stubs, use the new plugin, and change sql_utils.py:
```
from sqlalchemy import Mapped as Mapped
```

Other things I played around with were adding

```
if TYPE_CHECKING:
    Base = declarative_base()
else:
    from sql_utils.my_base import Base
    
 class MyTable(Base):
```

Thinking maybe I am getting all these errors due to the original issue I asked about, and tried to follow this advice to trigger the plugin. Didn't work though, ended up with 11k errors looking like:

`has incompatible type "Column[str]"; expected "str"`

So I guess my two questions are:
- are all these type errors expected or does it indicate the plugin isn't working properly?
- Is it a best practice to add explicit types to each column (e.g. `: Mapped[bool]`)?

**Reply by zzzeek** [reply]
> Thank you for your continued help! I'll break down some of the 9k errors.
> 
> **Need type annotation for [a column]**
> 
> Show up for tables where the columns are defined like this:
> 
> ```
>     creation_order_id = Column(BigInteger().with_variant(Integer(), "sqlite"), primary_key=True, autoincrement=True)
>     foo_sid = Column(String(34), nullable=False)
>     is_bar = Column(Boolean, nullable=False)
> ```

I would change those to `Mapped[int]` right off and use `mapped_column()` directly.

For the String(34) and Boolean the mypy plugin is supposed to handle that.


> 
> **Incompatible types in assignment** seems to show up when I have a column with an explicit type, defined like this:
> 
> ```
>     status: Column[ALiteralStatusType] = Column(
>         Enum(*StringLiteralUtils.get_literal_type_values(ALiteralStatusType)), nullable=False
>     )
> ```

that's Dropbox's typing form, which is incorrect.    Usage of `Column[]` on the left side must be swapped out globally; also we would assume `ALiteralStatusType` is an `enum.Enum`, you can use that directly.  Using it as the plugin would accept it:

    status: ALiteralStatusType = Column(
         Enum(ALiteralStatusType), nullable=False
     )

using it in new style would be:

    status: Mapped[ALiteralStatusType] = mapped_column(
         Enum(ALiteralStatusType), nullable=False
     )

> 
> **has incompatible type "str | None"; expected "str"**
> 
> Seems to show up even though the column has nullable false. Perhaps would go away if I explicitly set the type.
> 
> ```
> my_id_column = Column(String(36), nullable=False, unique=True)
> ```

this is documented for the plugin here: https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html#introspection-of-columns-based-on-typeengine  the above column will be considered to be `Optional[str]` even though nullable=False.   You would put a type annotation here, which can be `Mapped[str]` that would work both with and without the plugin.

This program passes mypy 100% on SQLAlchemy 2.0 with the mypy plugin properly configured, see that you can get it all green:

```py
import enum

from sqlalchemy import BigInteger
from sqlalchemy import Boolean
from sqlalchemy import Column
from sqlalchemy import Enum
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column

Base = declarative_base()



class ALiteralStatusType(enum.Enum):
    x = "somestatus"

class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)

    creation_order_id: Mapped[int] = Column(BigInteger().with_variant(Integer(), "sqlite"), primary_key=True, autoincrement=True)
    my_id_column = Column(String(36), nullable=False, unique=True)
    status1: ALiteralStatusType = Column(
         Enum(ALiteralStatusType), nullable=False
     )
    status2: Mapped[ALiteralStatusType] = mapped_column(
         Enum(ALiteralStatusType), nullable=False
     )

    foo_sid = Column(String(34), nullable=False)
    is_bar = Column(Boolean, nullable=False)
```

**Comment by ghost** [comment]
I've got most call sites now migrated to `mapped_column` and `Mapped[{type here}]` style and I'm down to about ~1588 errors.

Many of these are coming from our alembic migration files. 
Something like this:
`existing_server_default=sa.text("CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP"),`

will raise an error when I move off the stubs. Changing it to:

`existing_server_default=sa.text("CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP").text,`

seems to do the trick. However, it's concerning because these alembic files are auto-generated and I'm not keen on manually changing them. I also need to investigate to see whether the newer generated alembic files will use the correct variation to avoid future errors.

Is this an issue you've heard about before?

**Reply by CaselIT** [reply]
The bit about alembic may be a type bug there. If the latest versions also have it feel free to open a pr

**Reply by ghost** [reply]
I upgraded alembic to 1.13 and still got the error.  Do you think it's a bug in alembic or a bug in sql alchemy types system?

From a brief search on github, I don't see anyone with alembic files with that parameter suffixed with `sa.text(...).text`.

Do you think the alembic type just needs to be updated to accept `TextClause`?

**Reply by ghost** [reply]
Or maybe it's `TextClause` that needs to change, so it can be passed into functions expecting type `str`, since it has a `__str__` fn.

**Reply by CaselIT** [reply]
yes, that's a bug. In half the places server_defauly is typed with `Union["TextClause", "FetchedValue", "Function[Any]", str]` that's correct, while in others is typed as `Union[str, bool, Identity, Computed, None]`

Will open an issue -> tracked at https://github.com/sqlalchemy/alembic/issues/1468

**Comment by derekmartin3737-coder** [comment]
If part of the migration work here is still the repetitive 1.4-era runtime API cleanup, I built a narrow scanner-first tool for that side of the move.

It does not solve the typing or mypy-plugin part. What it does do is qualify a repo for supported legacy cleanup like `Query.get()`, `select([..])`, simple string joins and loader paths, and declarative import moves while keeping unsupported cases visible instead of pretending the whole migration is automated.

Overview:
https://zippertools.org/sqlalchemy-migration-tool.html
